core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -217,6 +253,112 @@ private void losePrimacy() throws Exception {
+          ProcessUtils.fatalError(LOG,
+                  "Failed to stop serving thread after %dms. Serving thread stack trace:%n%s",
+                  mServingThreadTimeoutMs, ThreadUtils.formatStackTrace(mServingThread));
+        }
+        mServingThread = null;
+      } catch (Exception e) {
+        ProcessUtils.fatalError(LOG, e, "Failed to stop serving thread during the restart. The process will exit.");
+      }
+    }
+    try {
+      stopMasters();
+      LOG.info("All master modules stopped");
+    } catch (Exception e) {
+      ProcessUtils.fatalError(LOG, e, "Failed to stop master modules during the restart. The process will exit.");
+    }
+
+
+    // Move the journal folder to a recycle bin and recreate a clean journal path
+    try {
+      String journalDir = Configuration.getString(PropertyKey.MASTER_JOURNAL_FOLDER);
+      String journalNewPath = journalDir + "_backup_" + new SimpleDateFormat("yyyyMMdd_hhmmss").format(new Date());
+      LOG.info("Moving journals from {} to {}", journalDir, journalNewPath);
+      FileUtils.move(journalDir, journalNewPath);
+      LOG.info("Journal files moved from {} to {}", journalDir, journalNewPath);
+      FileUtils.createDir(journalDir);
+      LOG.info("Recreated journal dir {}, accessible: {}", journalDir, FileUtils.isStorageDirAccessible(journalDir));
+    } catch (Exception e) {
+      ProcessUtils.fatalError(LOG, e, "Failed to backup and format journal directory during the restart. The process will exit.");
+    }
+
+    // Recreate the JournalSystem
+    URI journalLocation = JournalUtils.getJournalLocation();
+    JournalSystem newJournalSystem = new JournalSystem.Builder()
+            .setLocation(journalLocation).build(CommonUtils.ProcessType.MASTER);
+    try {
+      newJournalSystem.start();
+    } catch (Exception e) {
+      ProcessUtils.fatalError(LOG, e, "Failed to restart the raft journal. The process will quit and you will need to restart manually.");
+    }
+
+    // At the end, replace the reference with the new one
+    LOG.info("Switch the master to the fresh Ratis journal and update the runtime context.");
+    mJournalSystem = newJournalSystem;
+    // TODO(jiacheng): replace the reference in the CoreMasterContext
+    // Create masters.
+    String baseDir = Configuration.getString(PropertyKey.MASTER_METASTORE_DIR);
+    mContext = CoreMasterContext.newBuilder()
+            .setJournalSystem(mJournalSystem)
+            .setSafeModeManager(mSafeModeManager)
+            .setBackupManager(mBackupManager)
+            .setBlockStoreFactory(MasterUtils.getBlockStoreFactory(baseDir))
+            .setInodeStoreFactory(MasterUtils.getInodeStoreFactory(baseDir))
+            .setStartTimeMs(mStartTimeMs)
+            .setPort(NetworkAddressUtils
+                    .getPort(NetworkAddressUtils.ServiceType.MASTER_RPC, Configuration.global()))
+            .setUfsManager(mUfsManager)
+            .build();
+    MasterUtils.createMasters(mRegistry, mContext);
```

Comment:
The masters MUST be recreated using the new journal. It's just so hard to reuse the masters because that requires us to:
1. close cleanly
2. wipe out cleanly